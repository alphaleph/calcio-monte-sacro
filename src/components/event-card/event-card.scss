@import '../../pages/globals.scss';

.event-card-container {
    min-width: 280px;
    min-height: 280px;
    width: 30vw;
    height: 30vw;
    background-color: $black-main;
    color: $white-main;
    font-weight: bold;
}

.event-card-header {
    height: 20%;
    padding: .75rem;
}

.event-card-info {
    height: 80%;
    margin: 0;
}

.team-info {
    font-size: 1.5rem;
}

.match-info {
    font-size: 2.3rem;
    padding: 0rem;
}

.practice-info {
    font-size: 2.3rem;
}

@media screen and (min-width: 1023px) {
    .team-info {
        font-size: 2rem;
    }
    .match-info {
        font-size: 2.5rem;
    }
    .practice-info {
        font-size: 2.5rem;
    }
}