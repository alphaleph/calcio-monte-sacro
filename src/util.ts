import { MarkdownDataNode } from './types/index';

/**@func findMostRecentlyPastIndex
 * Finds index of most recently past Markdown post. Implemented via a modified binary search.
 * @param {MarkdownDataNode[]} edges Array of Markdown nodes. Assumes dates are sorted in ASC order!
 * @return {number} The index of the most recently past Markdown post. Returns -1 if current date is the earliest date.
 * @throws Error if edges size is 0.
 */
export const findMostRecentlyPastIndex = (
    edges: MarkdownDataNode[]
): number => {
    if (edges.length === 0) {
        throw Error('Error: There are no Markdown nodes for searching!');
    }

    const currDate = Date.now();

    // Case: Returns latest post if today is latest date
    // Case: Returns -1 if today is earliest date
    if (Date.parse(edges[edges.length - 1].node.frontmatter.date) < currDate) {
        return edges.length - 1;
    } else if (Date.parse(edges[0].node.frontmatter.date) > currDate) {
        return -1;
    }

    // Case: Return index of most recently past post
    const dates = edges.map((edge: MarkdownDataNode) =>
        Date.parse(edge.node.frontmatter.date)
    );

    return lowestNeighborIndexBinarySearch(dates, currDate);
};

/**@func lowestNeighborIndexBinarySearch
 * Finds via binary search: the index of the query value, of the nearest lower array value, or of the lowest array value if query value is the lowest.
 * @param {number[]} arr Array of numbers. Assumes dates are sorted in ASC order!
 * @param {number} val Value whose index to search for.
 * @return {number} The index of the query value. Otherwise, the index of the lowest neighbor.
 * @throws Error if array size is 0.
 */
export const lowestNeighborIndexBinarySearch = (
    arr: number[],
    val: number
): number => {
    if (arr.length === 0) {
        throw Error('Error: Array has size less than 1!');
    }

    let currIndex = Math.floor(arr.length / 2);
    let finalStep = Math.floor(Math.log2(arr.length)) - 1;
    let indexModifier = 0;

    for (let i = 0; i < Math.floor(Math.log2(arr.length)); i++) {
        if (arr[currIndex] === val) {
            return currIndex;
        }

        indexModifier = Math.max(
            Math.floor(Math.floor(arr.length / (2 * (i + 1))) / 2),
            1
        );

        if (arr[currIndex] < val) {
            //Look into upper half
            if (
                i < finalStep ||
                (arr.length % 2 === 0 && val >= arr[currIndex + indexModifier])
            ) {
                currIndex += indexModifier;
            }

            // Corner Case for two-off odd-sized arrays
            if (
                i === finalStep &&
                (arr.length % 2 === 1 && val >= arr[currIndex + indexModifier])
            ) {
                currIndex += indexModifier;
                // Inner Corner Case for one-off odd-sized arrays
                if (
                    currIndex !== arr.length &&
                    val >= arr[currIndex + indexModifier]
                ) {
                    currIndex += indexModifier;
                }
            }
        } else if (arr[currIndex] > val) {
            //Look into lower half
            currIndex -= indexModifier;
            // Corner Case due to one-off
            if (i === finalStep && (currIndex != 0 && val < arr[currIndex])) {
                currIndex -= indexModifier;
            }
        }
    }
    return currIndex;
};
