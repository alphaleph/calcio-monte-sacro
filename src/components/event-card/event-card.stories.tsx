import React from 'react';
import { storiesOf } from '@storybook/react';
import { EventCard } from './event-card';
import '../../pages/index.scss';
import {
    mockEventPracticeEN,
    mockEventPracticeIT,
    mockEventFinishedMatchEN,
    mockEventFinishedMatchIT,
    mockEventFutureMatchEN,
    mockEventFutureMatchIT,
} from '../__tests__/test-util';

storiesOf('EventCard', module)
    .add('practice-en', () => <EventCard event={mockEventPracticeEN} />)
    .add('match-finished-en', () => (
        <EventCard event={mockEventFinishedMatchEN} />
    ))
    .add('match-future-en', () => <EventCard event={mockEventFutureMatchEN} />)
    .add('practice-it', () => <EventCard event={mockEventPracticeIT} />)
    .add('match-finished-it', () => (
        <EventCard event={mockEventFinishedMatchIT} />
    ))
    .add('match-future-it', () => <EventCard event={mockEventFutureMatchIT} />)
    .add('empty', () => <EventCard event={null} />);
