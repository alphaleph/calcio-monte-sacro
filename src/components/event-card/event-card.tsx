import React from 'react';
import { Loading } from '../loading/loading';
import { MarkdownDataNode } from '../../types/index';
import { MATCH_EN, MATCH_IT, HOME_VENUE, TEAM_ABRV } from '../../constants';
import { getMonthAbrvEN } from '../../util';
import './event-card.scss';

interface EventCardProps {
    event: MarkdownDataNode | null;
}

export const EventCard = ({ event }: EventCardProps) => {
    let el = <Loading />;
    if (event) {
        const frontMatter = (event as MarkdownDataNode).node.frontmatter;
        const date = new Date(frontMatter.date);
        const venue = frontMatter.venue;
        const type = frontMatter.type;
        // Event: Match
        if (type === MATCH_EN || type === MATCH_IT) {
            const opposingTeam = frontMatter.opposing_team;
            let matchInfo = null;
            // Event: Finished Match
            if (date.getTime() < Date.now()) {
                const scoreMS = frontMatter.score_ms;
                const scoreOther = frontMatter.score_other;
                matchInfo = (
                    <p>{`${venue === HOME_VENUE ? scoreMS : scoreOther}-${
                        venue === HOME_VENUE ? scoreOther : scoreMS
                    }`}</p>
                );
            } else {
                // Event: Future Match
                matchInfo = <p>{`${date.getHours()}:${date.getMinutes()}`}</p>;
            }
            el = (
                <article className="event-card-container container">
                    <header className="event-card-header">
                        <p>
                            {`${date.getDate()} ${getMonthAbrvEN(
                                date.getMonth()
                            )} ${date.getFullYear()}`}
                        </p>
                        <p className="has-text-weight-light">{venue}</p>
                    </header>
                    <section className="event-card-info columns is-mobile is-vcentered has-text-centered">
                        <div className="column is-one-third team-info">
                            <p
                                className="is-size-1"
                                role="img"
                                aria-label="Soccer ball"
                            >
                                ⚽
                            </p>
                            <p>
                                {venue === HOME_VENUE
                                    ? TEAM_ABRV
                                    : opposingTeam}
                            </p>
                        </div>
                        <div className="column is-one-third match-info has-text-team-holy">
                            {matchInfo}
                        </div>
                        <div className="column is-one-third team-info">
                            <p
                                className="is-size-1"
                                role="img"
                                aria-label="Soccer ball"
                            >
                                ⚽
                            </p>
                            <p>
                                {venue === HOME_VENUE
                                    ? opposingTeam
                                    : TEAM_ABRV}
                            </p>
                        </div>
                    </section>
                </article>
            );
        } else {
            // Event: Practice
            el = (
                <article className="event-card-container container">
                    <header className="event-card-header">
                        <p>
                            {`${date.getDate()} ${getMonthAbrvEN(
                                date.getMonth()
                            )} ${date.getFullYear()}`}
                        </p>
                        <p className="has-text-weight-light">{venue}</p>
                    </header>
                    <section className="event-card-info has-text-centered columns is-mobile is-vcentered">
                        <div className="column practice-info">
                            <p>PRACTICE</p>
                            <p className="has-text-team-holy">{`${date.getHours()}:${date.getMinutes()}`}</p>
                        </div>
                    </section>
                </article>
            );
        }
    } else {
        // Event: null
        el = (
            <article
                className="event-card-container container"
                data-testid="event-card-empty"
            />
        );
    }

    return el;
};
