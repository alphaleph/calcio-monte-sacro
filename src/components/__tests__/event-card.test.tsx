import React from 'react';
import { EventCard } from '../event-card/event-card';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import {
    mockEventPracticeEN,
    mockEventFinishedMatchEN,
    mockEventFutureMatchEN,
} from '../__tests__/test-util';
import { getMonthAbrvEN } from '../../util';

const VENUE = 'Parco Sannazzaro';
const OPP_TEAM = 'CMS2';
const SCORE_MS = 1;
const SCORE_OTHER = 2;
const HOME_SCORE_STRING = `${SCORE_MS}-${SCORE_OTHER}`;
const rawDateFinished = new Date('2019-11-20T17:40:15.578Z');
const DATE_FINISHED = `${rawDateFinished.getDate()} ${getMonthAbrvEN(
    rawDateFinished.getMonth()
)} ${rawDateFinished.getFullYear()}`;
const rawDateFuture = new Date('2200-11-20T17:40:15.578Z');
const DATE_FUTURE = `${rawDateFuture.getDate()} ${getMonthAbrvEN(
    rawDateFuture.getMonth()
)} ${rawDateFuture.getFullYear()}`;
const TIME = `${rawDateFuture.getHours()}:${rawDateFuture.getMinutes()}`;
const EMPTY_TEST_ID = 'empty';

//TODO: Error checks for incorrectly formatted input
//TODO: Test cases for non-home venues
//TODO: Replace team emblem placeholders
describe('Event Card', () => {
    it('renders without crashing--practice', () => {
        const { getByText } = render(<EventCard event={mockEventPracticeEN} />);
        expect(getByText(VENUE)).toBeInTheDocument();
    });

    it('renders without crashing--match-finished', () => {
        const { getByText } = render(
            <EventCard event={mockEventFinishedMatchEN} />
        );
        expect(getByText(VENUE)).toBeInTheDocument();
    });

    it('renders without crashing--match-future', () => {
        const { getByText } = render(
            <EventCard event={mockEventFutureMatchEN} />
        );
        expect(getByText(VENUE)).toBeInTheDocument();
    });

    it('renders without crashing--null', () => {
        const { getByTestId } = render(<EventCard event={null} />);
        expect(getByTestId(EMPTY_TEST_ID)).toBeInTheDocument();
    });

    it('renders correctly--practice', () => {
        const { getByText } = render(<EventCard event={mockEventPracticeEN} />);
        expect(getByText(VENUE)).toBeVisible();
        expect(getByText(DATE_FINISHED)).toBeVisible();
        expect(getByText(TIME)).toBeVisible();
        expect(getByText('PRACTICE')).toBeVisible();
    });

    it('renders correctly--match-finished', () => {
        const { getByText } = render(
            <EventCard event={mockEventFinishedMatchEN} />
        );
        expect(getByText(VENUE)).toBeVisible();
        expect(getByText(DATE_FINISHED)).toBeVisible();
        expect(getByText(OPP_TEAM)).toBeVisible();
        expect(getByText(HOME_SCORE_STRING)).toBeVisible();
    });

    it('renders correctly--match-future', () => {
        const { getByText } = render(
            <EventCard event={mockEventFutureMatchEN} />
        );
        expect(getByText(VENUE)).toBeVisible();
        expect(getByText(DATE_FUTURE)).toBeVisible();
        expect(getByText(TIME)).toBeVisible();
        expect(getByText(OPP_TEAM)).toBeVisible();
    });

    it('renders correctly--null', () => {
        const { getByTestId } = render(<EventCard event={null} />);
        expect(getByTestId(EMPTY_TEST_ID)).toBeVisible();
    });

    it('has not changed--practice', () => {
        const { asFragment } = render(
            <EventCard event={mockEventPracticeEN} />
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it('has not changed--match-finished', () => {
        const { asFragment } = render(
            <EventCard event={mockEventFinishedMatchEN} />
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it('has not changed--match-future', () => {
        const { asFragment } = render(
            <EventCard event={mockEventFutureMatchEN} />
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it('has not changed--null', () => {
        const { asFragment } = render(<EventCard event={null} />);
        expect(asFragment()).toMatchSnapshot();
    });
});
