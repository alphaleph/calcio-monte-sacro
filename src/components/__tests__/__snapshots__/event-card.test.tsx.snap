// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Event Card has not changed--match-finished 1`] = `
<DocumentFragment>
  <article
    class="event-card-container container"
  >
    <header
      class="event-card-header"
    >
      <p>
        20 NOV 2019
      </p>
      <p
        class="has-text-weight-light"
      >
        Parco Sannazzaro
      </p>
    </header>
    <section
      class="event-card-info columns is-mobile is-vcentered has-text-centered"
    >
      <div
        class="column is-one-third team-info"
      >
        <p
          aria-label="Soccer ball"
          class="is-size-1"
          role="img"
        >
          ⚽
        </p>
        <p>
          CMS
        </p>
      </div>
      <div
        class="column is-one-third match-info has-text-team-holy"
      >
        <p>
          1-2
        </p>
      </div>
      <div
        class="column is-one-third team-info"
      >
        <p
          aria-label="Soccer ball"
          class="is-size-1"
          role="img"
        >
          ⚽
        </p>
        <p>
          CMS2
        </p>
      </div>
    </section>
  </article>
</DocumentFragment>
`;

exports[`Event Card has not changed--match-future 1`] = `
<DocumentFragment>
  <article
    class="event-card-container container"
  >
    <header
      class="event-card-header"
    >
      <p>
        20 NOV 2200
      </p>
      <p
        class="has-text-weight-light"
      >
        Parco Sannazzaro
      </p>
    </header>
    <section
      class="event-card-info columns is-mobile is-vcentered has-text-centered"
    >
      <div
        class="column is-one-third team-info"
      >
        <p
          aria-label="Soccer ball"
          class="is-size-1"
          role="img"
        >
          ⚽
        </p>
        <p>
          CMS
        </p>
      </div>
      <div
        class="column is-one-third match-info has-text-team-holy"
      >
        <p>
          12:40
        </p>
      </div>
      <div
        class="column is-one-third team-info"
      >
        <p
          aria-label="Soccer ball"
          class="is-size-1"
          role="img"
        >
          ⚽
        </p>
        <p>
          CMS2
        </p>
      </div>
    </section>
  </article>
</DocumentFragment>
`;

exports[`Event Card has not changed--null 1`] = `
<DocumentFragment>
  <article
    class="event-card-container container"
    data-testid="empty"
  />
</DocumentFragment>
`;

exports[`Event Card has not changed--practice 1`] = `
<DocumentFragment>
  <article
    class="event-card-container container"
  >
    <header
      class="event-card-header"
    >
      <p>
        20 NOV 2019
      </p>
      <p
        class="has-text-weight-light"
      >
        Parco Sannazzaro
      </p>
    </header>
    <section
      class="event-card-info has-text-centered columns is-mobile is-vcentered"
    >
      <div
        class="column practice-info"
      >
        <p>
          PRACTICE
        </p>
        <p
          class="has-text-team-holy"
        >
          12:40
        </p>
      </div>
    </section>
  </article>
</DocumentFragment>
`;
